db/document.go

```diff
@@ -106,6 +109,86 @@ func UnmarshalDocumentSyncData(data []byte, needHistory bool) (*syncData, error)
 	return root.SyncData, nil
 }
 
+// Unmarshals sync metadata for a document arriving via DCP.  Includes handling for xattr content
+// being included in data.  If not present in either xattr or document body, returns nil but no error.
+// We may need to
+func UnmarshalDocumentSyncDataFromFeed(data []byte, dataType uint8, needHistory bool) (result *syncData, err error) {
+
+	var body []byte
+
+	// If attr datatype flag is set, data includes both xattrs and document body.  Check for presence of sync xattr.
+	// Note that there could be a non-sync xattr present
+	if dataType&base.MemcachedDataTypeXattr != 0 {
+		var syncXattr []byte
+		body, syncXattr, err = parseXattrStreamData(KSyncXattrName, data)
+		if err != nil {
+			return nil, err
+		}
+
+		// If the sync xattr is present, use that to build syncData
+		if syncXattr != nil && len(syncXattr) > 0 {
+			result = &syncData{}
+			if needHistory {
+				result.History = make(RevTree)
+			}
+			err = json.Unmarshal(syncXattr, result)
+			if err != nil {
+				return nil, err
+			}
+			return result, nil
+		}
+	} else {
+		// Xattr flag not set - data is just the document body
+		body = data
+	}
+
+	// Non-xattr data, or sync xattr not present.  Attempt to retrieve sync metadata from document body
+	return UnmarshalDocumentSyncData(body, needHistory)
+
+}
+
+// parseXattrStreamData returns the raw bytes of the body and the requested xattr (when present) from the raw DCP data bytes.
+// Details on format: https://docs.google.com/document/d/18UVa5j8KyufnLLy29VObbWRtoBn9vs8pcxttuMt6rz8/edit#heading=h.caqiui1pmmmb
+func parseXattrStreamData(xattrName string, data []byte) (body []byte, xattr []byte, err error) {
+
+	xattrsLen := binary.BigEndian.Uint32(data[0:4])
+
+	body = data[xattrsLen:]
+	if xattrsLen == 0 {
+		return body, nil, nil
+	}
+
+	// In the xattr key/value pairs, key and value are both terminated by 0x00 (byte(0)).
+	separator := []byte{byte(0)}
+
+	// Iterate over xattrs
+	pos := uint32(4)
+	for pos < xattrsLen {
+		pairLen := binary.BigEndian.Uint32(data[pos : pos+4])
```

Comment:
I'm not convinced these changes make the code more understandable (as opposed to just more busy).  I've added some additional comments to make it clear that the code is processing xattr k/v pairs, though.